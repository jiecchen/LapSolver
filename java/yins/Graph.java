/**
 * @file   Graph.java
 * @author Daniel Spielman <spielman@cs.yale.edu>
 * @date   Sun Dec 22 2013
 *
 * (c) Daniel Spielman, 2013, part of the YINSmat package
 *
 * A sparse UNWEIGHTED static graph class.
 * keeps an adj list for each vertex, 
 * not necessarily sorted
 * produces edge back indices only when needed
 *
 * Meant for use inside matlab, but this causes 0/1 indexing problems!
 * Matlab indexes from 1, this code from 0.
 * Matlab must do all the conversions.
 *
 * To create in matlab, use
 * [i,j,v] = find(tril(T))
 * graph = WeightedGraph(i-1,j-1,v)
 *
 * Elementary methods: dfs, components, bfsWalk, 
 *
 */

package yins;

import java.lang.*;
import java.util.*;
import java.io.*;


public class Graph 
{

    //-----------------------
    //   the number of verts
    //
    public int nv;

    //---------------------
    //  the main graph data,
    //  nbrs[x][i] is ith nbr 
    //       of node x
    //
    //  deg[x] is the degree of node x

    public int[][] nbrs;
    public int[] deg;


    
    //-----------------------------
    //  nbrs[nbrs[x][i],backInd[x][i]] = x
    public int[][] backInd;
    
    //--------------------
    //  a dfs ordering
    //  not initialized
    //
    private int[] dfs;

    private int dfsPtr; // a utility


    //--------------------
    //  a bfs ordering,
    //  and bfs depth
    //
    public int[] bfs;
    public int[] depth;


    //------------------------
    //  the parent in the dfs tree
    //  initialized by dfs()
    //
    private int[] parent;
    private int root;

    public boolean[] seen;
    public int[] comp;


    public Graph (int[] i, int[] j) {
        this.setGraph(i,j);
    }


    
    /**
     *  expects input as
     *  [i,j] = find(tril(T)),
     *  so, each entry of i should be larger than corresp entry of j
     */
    public void setGraph (int[] i, int[] j) {
        int len; // the length of i

        dfs = null;
        comp = null;
        seen = null;

        
        len = i.length;
        if ((j.length != len)) {
            throw new Error("inputs must all have the same length");
        }

        //-------------------------
        // compute max node index
        //
        this.nv = 0;
        for (int a = 0; a < len; a++) {
            if (i[a] > nv)
                nv = i[a];
            if (j[a] > nv)
                nv = j[a];
        }


        //-------------------------
        // downshift i and j by 1
        // 
        for (int a = 0; a < len; a++) {
            i[a] = i[a] - 1;
            j[a] = j[a] - 1;

            // report error if self-loops
            if (i[a] == j[a]) {
                throw new Error("Self-loops are not allowed.");
            }
        }


        //-----------------------------------------
        //  count how many times each node occurrs
        
        deg = new int[nv];
        for (int a = 0; a < nv; a++) {
            deg[a] = 0;
        }

        for (int a = 0; a < len; a++) {
            if (i[a] > j[a]) {
                deg[i[a]] = deg[i[a]] + 1;
                deg[j[a]] = deg[j[a]] + 1;
            }
        }


        //---------------------------
        // build the graph

        int[] tmpdeg = new int[nv];

        nbrs = new int[nv][];

        for (int a = 0; a < nv; a++) {
            nbrs[a] = new int[deg[a]];
            tmpdeg[a] = 0;
        }

        for (int a = 0; a < len; a++) {
            if (i[a] > j[a]) {
            
                // backInd[i[a]][tmpdeg[i[a]]] = tmpdeg[j[a]];
                // backInd[j[a]][tmpdeg[j[a]]] = tmpdeg[i[a]];
                
                nbrs[i[a]][tmpdeg[i[a]]++] = j[a];
                nbrs[j[a]][tmpdeg[j[a]]++] = i[a];
            }
        }
    }

    /*
     * The back indices satisfy
     * nbrs[nbrs[x][i],backInd[x][i]] = x
     *
     */
    public void makeBackEdges() {
        backInd = new int[nv][];

        int[] count = new int[nv];
        
        for (int a = 0; a < nv; a++) {
            backInd[a] = new int[deg[a]];
            count[a] = 0;
        }

        for (int a = 0; a < nv; a++) {
            for (int i = 0; i < deg[a]; i++) {
                int nbr = nbrs[a][i];
                backInd[nbr][count[nbr]++] = i;
            }
        }
    }
    
    /** 
     * walk in BFS order, setting the field bfs
     * assumes connected
     */
    public int[] bfsWalk(int root) {

        bfs = new int[nv];
        depth = new int[nv];
        seen = new boolean[nv];

        for (int i = 0; i < nv; i++) {
            seen[i] = false;
        }
        
        depth[root] = 0;
        bfs[0] = root;
        seen[root] = true;
        
        int bfsPtr = 1;
        int curPtr = 1;
        int curNode = root;

        while (curPtr < nv) {
            int nbr;
            for (int i = 0; i < deg[curNode]; i++) {
                nbr = nbrs[curNode][i];
                if (!seen[nbr]) {
                    bfs[bfsPtr++] = nbr;
                    depth[nbr] = depth[curNode]+1;
                    seen[nbr] = true;
                }
            }
            curNode = bfs[curPtr++];
        }

        return bfs;
    }


    /**
     * Compute dfs by a recursive algorithm
     * could easily break the stack
     */
    public int[] dfs () {

        int c = 0;
        
        seen = new boolean[nv];
        comp = new int[nv];
        for (int i = 0; i < nv; i++) {
            seen[i] = false;
            comp[i] = 0;
        }
            
        dfsPtr = 0;
        dfs = new int[nv];
        parent = new int[nv];

        for (int x = 0; x < nv; x++) {
            if (!seen[x]) {
                dfsSub(x,c);
                c = c + 1;
            }
        }
        
        return dfs;
    }

    private void dfsSub(int x, int c) {
        dfs[dfsPtr++] = x;
        seen[x] = true;
        comp[x] = c;
        for (int i = 0; i < deg[x]; i++) 
            if (!seen[nbrs[x][i]]) {
                parent[nbrs[x][i]] = x;
                dfsSub(nbrs[x][i],c);
            }

    }
        
    public int[] getDfs () {return dfs;}


    /**
     * Compute connected components using a bfs approach.
     * Returns a the characteristic vector of the components,
     * starting to index with 1
     */
    public int[] components () {

        int[] order = new int[nv];

        seen = new boolean[nv];
        comp = new int[nv];
        
        for (int i = 0; i < nv; i++) {
            comp[i] = 0;
        }


        int c = 0;
        for (int x = 0; x < nv; x++) {
            if (comp[x] == 0) {
                c = c + 1;
                comp[x] = c;
                int ptr = 0;
                int orderLen = 1;
                order[ptr] = x;
                
                while (ptr < orderLen) {
                    int curNode = order[ptr];
                    for (int i = 0; i < deg[curNode]; i++) {
                        int nbr = nbrs[curNode][i];
                        if (comp[nbr] == 0) {
                            comp[nbr] = c;
                            order[orderLen++] = nbr;
                        }
                    }
                    ptr++;
                }
            }
        }
        return comp;
    }

    
    
    /* if the graph is a tree,
       this turns it into a parent array
    */

    public int[] treeToArray() {

        dfs();
        
        int[] pArray = new int[nv];

        // set all others to parent
        for (int i = 0; i < nv; i++)
            pArray[i] = parent[i];

        // set root to itself
        pArray[dfs[0]] = dfs[0];

        return pArray;
        
    }



    /*
     * Find a maximal indep set by first choosing
     * vertices of low degree
     *
     * return the char vector of the set
     */
    public int[] bigIndepSet() {
        int[] charv = new int[nv];

        // 1 if in mis, 0 if not, -1 if unassigned
        for (int i = 0; i < nv; i++) {
            charv[i] = -1;
        }


        // form link lists for every degree, by ints
        // -1 if empty
        // list[v] is the next item after v
        int[] list = new int[nv];
        for (int i = 0; i < nv; i++) {
            list[i] = -1;
        }

        // also, keep a head of each list
        // -1 if empty
        int[] head = new int[nv];
        for (int i = 0; i < nv; i++) {
            head[i] = -1;
        }

        for (int v = 0; v < nv; v++) {
            int d = deg[v];
            list[v] = head[d];
            head[d] = v;
        }


        // now, start making the mis
        for (int d = 0; d < nv; d++) {
            int v = head[d];
            while (v != -1) {
                if (charv[v] == -1) {
                    charv[v] = 1;
                    for (int i = 0; i < deg[v]; i++) {
                        charv[nbrs[v][i]] = 0;
                    }
                }
                v = list[v];
            }
        }
        
        return charv;
    }

    

    public void dump() {

        for (int x = 0 ; x < nv; x++) {
            System.out.print(x +" : ");
            for (int i = 0; i < deg[x]; i++) 
                System.out.print(nbrs[x][i] + " ");
            System.out.println();
        }
        
    }


}
