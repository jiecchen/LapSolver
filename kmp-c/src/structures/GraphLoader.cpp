#include "GraphLoader.h"
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

Graph GraphLoader::fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w)
{
    return Graph(EdgeList(u, v, w));
}

Graph GraphLoader::fromFile(const string &filename)
{
    aligned_vector<int> u = aligned_vector<int>();
    aligned_vector<int> v = aligned_vector<int>();
    aligned_vector<double> w = aligned_vector<double>();

    string line;
    ifstream ijvFile(filename);
    if (!ijvFile)
        throw -1;

    while (getline(ijvFile, line))
    {
        if (line.empty()) break;
        istringstream iss(line);
        int a, b;
        double c;
        if (!(iss >> a >> b >> c))
            throw u.size();
        u.push_back(a);
        v.push_back(b);
        w.push_back(c);
    }

    return fromArrays(u, v, w);
}