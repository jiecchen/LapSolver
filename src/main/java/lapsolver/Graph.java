/**
 * @file Graph.java
 * @author Daniel Spielman <spielman@cs.yale.edu>
 * @author Cyril Zhang <cyril.zhang@yale.edu>
 * @date Sun Dec 22 2013
 *
 * (c) Daniel Spielman, 2013, part of the YINSmat package
 *
 * A sparse weighted static graph class.
 * keeps an adj list for each vertex, 
 * not necessarily sorted
 * produces edge back indices only when needed
 *
 * Meant for use inside matlab, but this causes 0/1 indexing problems!
 * Matlab indexes from 1, this code from 0.
 * Matlab must do all the conversions.
 *
 * To create in matlab, use
 * [i,j,v] = find(tril(T))
 * graph = Graph(i-1,j-1,v)
 *
 * Elementary methods: dfs, components, bfsWalk, 
 *
 */

package lapsolver;

public class Graph {
    public int nv; // number of vertices
    public int ne; // number of edges

    //---------------------
    //  the main graph data,
    //  nbrs[x][i] is ith nbr 
    //       of node x
    //  weights[x][i] is the weight of this edge
    //     (note each appears twice)
    //
    //  deg[x] is the degree of node x
    public int[][] nbrs;
    public double[][] weights;
    public int[] deg;

    //-----------------------------
    //  backInd[x][i] = the position of x in the neighbor list of u = graph.nbrs[x][i]
    //  nbrs[nbrs[x][i],backInd[x][i]] = x
    public int[][] backInd;

    public Graph() { }

    // constructor from edge list
    public Graph(EdgeList edges) {
        this(edges.u, edges.v, edges.weight);
    }

    // constructor from edge data
    public Graph(int[] src, int[] dst, double[] weight) {
        ne = src.length; // the length of i

        if ((dst.length != ne) || (weight.length != ne)) {
            throw new Error("inputs must all have the same length");
        }

        // Ensure src[i] > dst[i], src[i] != dst[i] \forall i
        for (int i = 0; i < ne; i++) {
            int lower = Math.min(src[i], dst[i]);
            int upper = Math.max(src[i], dst[i]);

            dst[i] = lower;
            src[i] = upper;

            if (src[i] == dst[i]) {
                throw new Error("Self-loops are not allowed (" + i + "/" + ne + ").");
            }
        }

        // compute max node auxiliarySize
        nv = 0;
        for (int i = 0; i < ne; i++) {
            if (src[i] > nv)
                nv = src[i];
            if (dst[i] > nv)
                nv = dst[i];
        }
        nv++; // largest 0-based auxiliarySize is nv-1

        //  count how many times each node occurs
        deg = new int[nv];
        for (int i = 0; i < nv; i++) {
            deg[i] = 0;
        }

        for (int i = 0; i < ne; i++) {
            deg[src[i]]++;
            deg[dst[i]]++;
        }

        // build the graph
        int[] tmpdeg = new int[nv];

        nbrs = new int[nv][];
        backInd = new int[nv][];
        weights = new double[nv][];

        for (int i = 0; i < nv; i++) {
            nbrs[i] = new int[deg[i]];
            backInd[i] = new int[deg[i]];
            weights[i] = new double[deg[i]];
            tmpdeg[i] = 0;
        }

        for (int i = 0; i < ne; i++) {
            if (src[i] > dst[i]) {
                weights[src[i]][tmpdeg[src[i]]] = weight[i];
                weights[dst[i]][tmpdeg[dst[i]]] = weight[i];

                backInd[src[i]][tmpdeg[src[i]]] = tmpdeg[dst[i]];
                backInd[dst[i]][tmpdeg[dst[i]]] = tmpdeg[src[i]];

                nbrs[src[i]][tmpdeg[src[i]]++] = dst[i];
                nbrs[dst[i]][tmpdeg[dst[i]]++] = src[i];
            }
        }
    }

    // get graph from tree
    public Graph(Tree tree) {
        this(new EdgeList(tree));
    }

    // copy constructor (perform a deep copy)
    public Graph(Graph other) {
        nv = other.nv;
        ne = other.ne;
        deg = other.deg.clone();

        nbrs = new int[nv][];
        weights = new double[nv][];
        backInd = new int[nv][];

        for (int i = 0; i < nv; i++) {
            nbrs[i] = other.nbrs[i].clone();
            weights[i] = other.weights[i].clone();
            backInd[i] = other.backInd[i].clone();
        }
    }
}
