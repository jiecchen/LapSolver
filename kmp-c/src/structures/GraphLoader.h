#pragma once

#include <string>
#include <iostream>
#include "structures/Graph.h"
#include "util/aligned.h"

class GraphLoader
{
public:
	
	static Graph fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w);
	static Graph fromFile(const std::string &filename);
    static Graph fromStream(std::istream &inStream);
    static Graph fromStdin();

};