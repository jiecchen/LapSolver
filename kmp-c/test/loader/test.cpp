// Test for GraphLoader

#include <cstdlib>
#include <string>
#include <structures/Graph.h>
#include <structures/GraphLoader.h>

int main(int argc, char const *argv[])
{
    atexit(mkl_free_buffers);
    Graph g = GraphLoader::fromStdin();

    printf("%d vertices\n", g.nv);
    for (int i = 0; i < g.nv; i++) {
        printf("deg(%d) = %d\n", i, g.getDegree(i));
    }

    return 0;
}
