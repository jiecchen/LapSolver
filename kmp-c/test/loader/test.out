26 vertices
deg(0) = 3
deg(1) = 3
deg(2) = 5
deg(3) = 4
deg(4) = 3
deg(5) = 5
deg(6) = 3
deg(7) = 2
deg(8) = 3
deg(9) = 2
deg(10) = 2
deg(11) = 2
deg(12) = 2
deg(13) = 1
deg(14) = 4
deg(15) = 1
deg(16) = 1
deg(17) = 3
deg(18) = 1
deg(19) = 1
deg(20) = 4
deg(21) = 1
deg(22) = 1
deg(23) = 1
deg(24) = 2
deg(25) = 2
